### ISSUE TITLE
Kubectl apply sometimes merges different volume types

### ISSUE THREAD

**Issue by nyarly** [kubernetes/kubernetes#112014]
### What happened?

I applied a new configuration to a series of statefulsets that changed a configMap volume into a projected volume. Some, but not all, of the statefulsets merged the volume definition so that the volume had two types, and later when the pods were replaced, they failed because `Forbidden: may not specify more than 1 volume type`

### What did you expect to happen?

The new configuration as a applied, with the `projected` type would replace the `configMap` type.

### How can we reproduce it (as minimally and precisely as possible)?

`kubectl create` with `configMap` volumes; update the config to `projected` and `apply.` 

### Anything else we need to know?

From what I can tell, it may not happen every time - I deployed 15 near-identical statefulsets in one `apply`, and only some of them wound up in the invalid state.

I see this comment https://github.com/kubernetes/kubernetes/issues/48285#issuecomment-318862830 on an old issue, which seems have rotted without address. The suggestion there is that this is an issue for SIG api-machinery?

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here

Client Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.7", GitCommit:"42c05a547468804b2053ecf60a3bd15560362fc2", GitTreeState:"archive", BuildDate:"1980-01-01T00:00:00Z", GoVersion:"go1.17.10", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.13-gke.900", GitCommit:"8f8dd9d389916c66dd26704a96f1ab4c5776f9df", GitTreeState:"clean", BuildDate:"2022-06-06T09:24:36Z", GoVersion:"go1.16.15b7", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>
GKE
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
   1   │ BUG_REPORT_URL="https://github.com/NixOS/nixpkgs/issues"
   2   │ BUILD_ID="22.05.20220811.45c9736"
   3   │ DOCUMENTATION_URL="https://nixos.org/learn.html"
   4   │ HOME_URL="https://nixos.org/"
   5   │ ID=nixos
   6   │ LOGO="nix-snowflake"
   7   │ NAME=NixOS
   8   │ PRETTY_NAME="NixOS 22.05 (Quokka)"
   9   │ SUPPORT_URL="https://nixos.org/community.html"
  10   │ VERSION="22.05 (Quokka)"
  11   │ VERSION_CODENAME=quokka
  12   │ VERSION_ID="22.05"
$ uname -a
# paste output here
Linux neumann 5.10.126 #1-NixOS SMP Mon Jun 27 07:41:01 UTC 2022 x86_64 GNU/Linux
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by nyarly** [comment]
/sig api-machinery

I think, anyway

**Comment by nyarly** [comment]
During recovery, it seems like the problem impacts older instances of the statefulset - I suspect that the difference is that the involved resources were originally created and updated with client-side apply, and we've since switched to doing SSA.

So, refined repro: 
* Start with statefulset.yaml, with a configmap volume
* `kubectl apply`
* update e.g. annotations, 
* `kubectl apply`
* change volume to projected
* `kubectl apply --server-side=true --field-manager=test`
* restart the statefulset

**Comment by liggitt** [comment]
cc @apelisse if the client-side → server-side apply transition is involved

**Comment by apelisse** [comment]
Yeah, we're working toward improving the client-side to server-side process.
In the mean-time, the workaround is to apply null for the former volume-type to force-remove it.

**Comment by nyarly** [comment]
/remove-lifecycle stale

I'm continuing to have weird merges during deploys

**Comment by apelisse** [comment]
More details would be useful. What "weird merges" do you have?

**Comment by nyarly** [comment]
See the initial report. I've noticed other cases where an apply has gone wrong, but it's been a couple of weeks since the last one.

**Comment by apelisse** [comment]
OK Thanks, I was curious if this was the same thing.

**Comment by gangstead** [comment]
I ran into this problem when I was changing a volume type (from using a PVC to an ephemeral volume).  The work around I made was to rename the volume (both in the declaration in `volumes` and where it is used in the container spec's  `volumeMounts`).  This way the controller doesn't try to merge the old and new definitions together.
